src/Workspaces/Remote/ServiceHub/Services/CodeAnalysisService_CodeLens.cs

```diff
@@ -63,12 +87,140 @@ public Task<string> GetFullyQualifiedName(DocumentId documentId, TextSpan textSp
+                // we have a lock here so that subscription and unsubscription of the two events
+                // are happening atomic, but that doesn't mean there is no possiblity of race here
+                // theoradically, there can be a race if connection got disconnected before we subscribe
+                // to OnRpcDisconnected but already in the subscription code path.
+                // but there is no easy way to solve the problem unless Rpc itself provide things like
+                // subscribe only if connection still alive or something
+                lock (_gate)
+                {
+                    if (subscription)
+                    {
+                        _owner.Disconnected += OnDisconnected;
+                        _workspace.WorkspaceChanged += OnWorkspaceChanged;
+                    }
+                    else
+                    {
+                        _owner.Disconnected -= OnDisconnected;
+                        _workspace.WorkspaceChanged -= OnWorkspaceChanged;
+                    }
+                }
+            }
+
+            private void OnDisconnected(object sender, EventArgs e)
+            {
+                ConnectEvents(subscription: false);
+            }
+
+            private void OnWorkspaceChanged(object sender, WorkspaceChangeEventArgs e)
+            {
+                EnqueueUpdate();
+                return;
+
+                void EnqueueUpdate()
+                {
+                    // workspace event is serialized events. and reset delay only get updated here
+                    if (!_resettableDelay.Task.IsCompleted)
+                    {
+                        _resettableDelay.Reset();
+                        return;
+                    }
+
+                    var delay = new ResettableDelay((int)s_delay.TotalMilliseconds);
+
+                    _resettableDelay = delay;
+                    delay.Task.ContinueWith(async _ =>
+                    {
+                        try
+                        {
+                            var document = _workspace.CurrentSolution.GetDocument(_documentId);
+                            if (document == null)
+                            {
+                                return;
+                            }
+
+                            var newVersion = await document.Project.GetDependentVersionAsync(CancellationToken.None).ConfigureAwait(false);
+                            if (newVersion == _lastVersion)
+                            {
+                                return;
+                            }
+
+                            // fire and forget.
+                            // ignore any exception such as rpc already disposed (disconnected)
+
+                            _lastVersion = newVersion;
+                            await _owner.InvokeAsync(nameof(IRemoteCodeLensDataPoint.Invalidate), CancellationToken.None).ConfigureAwait(false);
+                        }
+                        catch { }
```

Comment:
📝 This try/catch is unnecessary. The `ContinueWith` call does not restrict continuations to antecedent tasks with a specific final state, and the parameter is a discard.

Reply:
sure, I guess since body looks into exception of this task, just leaving it will make it to go to void :)